### PR TITLE
Using EIP-1559 V2 for swaps

### LINKED ISSUE #12597: Changes in swaps
Ensure following changes in swaps:

1. new confirm-transaction changes
2. new edit gas fee / advance gas inputs modals should be used

Screenshot 2021-11-05 at 2 29 24 PM

### PR DESCRIPTION
**Pull request by jpuri**
Fixes: #12597

Using 1559 V2 for swaps

https://user-images.githubusercontent.com/2182307/145007550-1ca377c2-c8d1-4ee4-ad54-360ca1f5bbfc.mov

### CODE REVIEW

File: ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js [hunk 1]
```diff
@@ -164,8 +165,13 @@ const EditGasItem = ({ priorityLevel }) => {
             <EditGasToolTip
               t={t}
               priorityLevel={priorityLevel}
-              maxFeePerGas={maxFeePerGas}
-              maxPriorityFeePerGas={maxPriorityFeePerGas}
+              maxFeePerGas={maxFeePerGas || maxFeePerGasValue}
```
Comment (mcmire) [inline ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js:168]: What is the difference between these two values? Is there a better name that we can give that would designate where these come from? I can see this being confusing.

File: ui/components/app/edit-gas-fee-popover/edit-gas-tooltip/edit-gas-tooltip.js [hunk 2]
```diff
@@ -64,10 +65,15 @@ const EditGasToolTip = ({
   return (
     <div className="edit-gas-tooltip__container">
       {priorityLevel !== PRIORITY_LEVELS.CUSTOM &&
```
Comment (mcmire) [inline ui/components/app/edit-gas-fee-popover/edit-gas-tooltip/edit-gas-tooltip.js:67]: Maybe not necessary for this PR, but I'm wondering if it's worth it at some point to extract these conditionals to a variable or refactor this in some way. It seems like a lot is going on here. Just a thought.

**Review by mcmire — DISMISSED** [review]
A few nits, but overall looks good to me.

File: ui/components/app/edit-gas-fee-button/edit-gas-fee-button.js [hunk 3]
```diff
@@ -28,13 +33,23 @@ export default function EditGasFeeButton({ userAcknowledgedGasMissing }) {
...
+  let icon = estimateUsed;
+  let title = estimateUsed;
+  if (
+    estimateUsed === PRIORITY_LEVELS.HIGH &&
+    editGasMode === EDIT_GAS_MODES.SWAPS
+  ) {
+    icon = 'swapSuggested';
+    title = 'swapSuggested';
+  }
+
```
Comment (adonesky1) [inline ui/components/app/edit-gas-fee-button/edit-gas-fee-button.js:45]: why not:
```suggestion
  if (
    estimateUsed === PRIORITY_LEVELS.HIGH &&
    editGasMode === EDIT_GAS_MODES.SWAPS
  ) {
    estimateUsed = 'swapSuggested';
  }

```
and then swap back `estimateUsed` for `icon` and `title` below

File: ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.js [hunk 4]
```diff
@@ -40,17 +43,23 @@ const EditGasFeePopover = () => {
                 <I18nValue messageKey="gasOption" />
               </span>
               <span className="edit-gas-fee-popover__content__header-time">
-                <I18nValue messageKey="time" />
+                {editGasMode !== EDIT_GAS_MODES.SWAPS && (
+                  <I18nValue messageKey="time" />
+                )}
```
Comment (adonesky1) [inline ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.js:48]: just curious why don't we include time estimations for swaps?

File: ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.js [hunk 5]
```diff
@@ -40,17 +43,23 @@ const EditGasFeePopover = () => {
...
               <span className="edit-gas-fee-popover__content__header-max-fee">
                 <I18nValue messageKey="maxFee" />
               </span>
             </div>
-            <EditGasItem priorityLevel={PRIORITY_LEVELS.LOW} />
+            {editGasMode !== EDIT_GAS_MODES.SWAPS && (
+              <EditGasItem priorityLevel={PRIORITY_LEVELS.LOW} />
+            )}
             <EditGasItem priorityLevel={PRIORITY_LEVELS.MEDIUM} />
             <EditGasItem priorityLevel={PRIORITY_LEVELS.HIGH} />
```
Comment (adonesky1) [inline ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.js:58]: Screen Shot 2021-12-08 at 3 27 51 PM
are we supposed to see the high/aggressive option?

Comment (adonesky1) [reply ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.js:58]: Ok I see [here](https://github.com/MetaMask/metamask-extension/pull/12966/files#diff-1a63f62a59b5000695c3aa2612b419287852e21c9dc2857e12c166ad75b99168R120) the test states we shouldn't. Why aren't we doing this filtering at this level?

File: ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.test.js [hunk 6]
```diff
@@ -103,12 +106,40 @@ describe('EditGasFeePopover', () => {
...
+    expect(screen.queryByText('⚙')).toBeInTheDocument();
+    expect(screen.queryByText('Low')).not.toBeInTheDocument();
+    expect(screen.queryByText('Market')).toBeInTheDocument();
+    expect(screen.queryByText('Aggressive')).not.toBeInTheDocument();
+    expect(screen.queryByText('Site')).not.toBeInTheDocument();
+    expect(screen.queryByText('Swap suggested')).toBeInTheDocument();
+    expect(screen.queryByText('Advanced')).toBeInTheDocument();
+  });
+
+  it('should not show time estimates fpr swaps', () => {
```
Comment (adonesky1) [inline ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.test.js:138]:
```suggestion
  it('should not show time estimates for swaps', () => {
```

File: ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js [hunk 7]
```diff
@@ -147,14 +158,20 @@ const EditGasItem = ({ priorityLevel }) => {
           '--'
         )}
       </span>
+      {/* todo: tooltip in case of swap suggested */}
```
Comment (adonesky1) [inline ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js:161]: I'm guessing we shouldn't consider this a blocker to this PR?

File: ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.test.js [hunk 8]
```diff
@@ -100,29 +107,41 @@ describe('EditGasItem', () => {
...
-    renderComponent({ priorityLevel: 'high' });
+    renderComponent({ componentProps: { priorityLevel: 'high' } });
     expect(screen.queryByRole('button', { name: 'high' })).toBeInTheDocument();
     expect(screen.queryByText('🦍')).toBeInTheDocument();
     expect(screen.queryByText('Aggressive')).toBeInTheDocument();
     expect(screen.queryByText('15 sec')).toBeInTheDocument();
     expect(screen.queryByTitle('0.0021 ETH')).toBeInTheDocument();
   });
 
+  it('should renders priorityLevel high as swap suggested for swaps', () => {
```
Comment (adonesky1) [inline ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.test.js:118]:
```suggestion
  it('should render priorityLevel high as swap suggested for swaps', () => {
```

File: ui/hooks/gasFeeInput/test-utils.js [hunk 9]
```diff
@@ -122,6 +126,12 @@ export const generateUseSelectorRouter = ({
       balance: '0x440aa47cc2556',
     };
   }
+  if (selector === getCustomMaxFeePerGas) {
+    return '0x5208';
+  }
+  if (selector === getCustomMaxPriorityFeePerGas) {
+    return '0x5208';
+  }
```
Comment (adonesky1) [inline ui/hooks/gasFeeInput/test-utils.js:134]: [nit]
```suggestion
  if (selector === getCustomMaxFeePerGas || selector === getCustomMaxPriorityFeePerGas) {
    return '0x5208';
  }
```

File: ui/hooks/gasFeeInput/useMaxFeePerGasInput.js [hunk 10]
```diff
@@ -34,14 +38,17 @@ const getMaxFeePerGasFromTransaction = (transaction) => {
...
+  editGasMode,
   estimateToUse,
   gasEstimateType,
   gasFeeEstimates,
   gasLimit,
   gasPrice,
+  supportsEIP1559V2,
   transaction,
 }) {
+  const swapCustomMaxFeePerGas = useSelector(getCustomMaxFeePerGas);
```
Comment (adonesky1) [inline ui/hooks/gasFeeInput/useMaxFeePerGasInput.js:50]: Not following why this value comes from customMaxFeePerGas?

File: ui/hooks/gasFeeInput/useMaxFeePerGasInput.js [hunk 11]
```diff
@@ -53,24 +60,31 @@ export function useMaxFeePerGasInput({
...
   const showFiat = useSelector(getShouldShowFiat);
 
-  const maxFeePerGasFromTransaction = supportsEIP1559
-    ? getMaxFeePerGasFromTransaction(transaction)
-    : 0;
+  let initialMaxFeePerGas;
+  if (editGasMode === EDIT_GAS_MODES.SWAPS && supportsEIP1559V2) {
+    initialMaxFeePerGas = swapCustomMaxFeePerGas
+      ? Number(hexWEIToDecGWEI(swapCustomMaxFeePerGas))
+      : null;
```
Comment (adonesky1) [inline ui/hooks/gasFeeInput/useMaxFeePerGasInput.js:67]: Should this be `0` instead of null?

File: ui/hooks/gasFeeInput/useMaxFeePerGasInput.js [hunk 12]
```diff
@@ -117,7 +131,7 @@ export function useMaxFeePerGasInput({
       gasFeeEstimates,
       gasEstimateType,
       estimateToUse,
-      maxFeePerGasFromTransaction,
+      initialMaxFeePerGas || 0,
```
Comment (adonesky1) [inline ui/hooks/gasFeeInput/useMaxFeePerGasInput.js:134]: and then this could just be `initialMaxFeePerGas` without `|| 0`

File: ui/hooks/gasFeeInput/useMaxPriorityFeePerGasInput.js [hunk 13]
```diff
@@ -52,22 +59,29 @@ export function useMaxPriorityFeePerGasInput({
...
   const showFiat = useSelector(getShouldShowFiat);
 
-  const maxPriorityFeePerGasFromTransaction = supportsEIP1559
-    ? getMaxPriorityFeePerGasFromTransaction(transaction)
-    : 0;
+  let initialMaxMaxPriorityFeePerGas;
+  if (editGasMode === EDIT_GAS_MODES.SWAPS && supportsEIP1559V2) {
+    initialMaxMaxPriorityFeePerGas = swapCustomMaxPriorityFeePerGas
+      ? Number(hexWEIToDecGWEI(swapCustomMaxPriorityFeePerGas))
+      : null;
```
Comment (adonesky1) [inline ui/hooks/gasFeeInput/useMaxPriorityFeePerGasInput.js:66]: also here? 0 instead of null?

File: ui/pages/confirm-transaction-base/gas-details-item/gas-details-item-title/gas-details-item-title.js [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+          <>
+            <Typography fontSize="12px">
+              {t('transactionDetailGasTooltipIntro', [
+                isMainnet ? t('networkNameEthereum') : '',
+              ])}
+            </Typography>
+            <Typography fontSize="12px">
+              {t('transactionDetailGasTooltipExplanation')}
+            </Typography>
+            <Typography fontSize="12px">
```
Comment (adonesky1) [inline ui/pages/confirm-transaction-base/gas-details-item/gas-details-item-title/gas-details-item-title.js:34]: Should we use the `variant` prop on these `Typography` components?

File: ui/pages/swaps/view-quote/view-quote.js [hunk 15]
```diff
@@ -672,93 +680,147 @@ export default function ViewQuote() {
...
+    userFeeLevel: swapsUserFeeLevel || GAS_RECOMMENDATIONS.HIGH,
+    txParams: {
+      maxFeePerGas,
+      maxPriorityFeePerGas,
+      gas: maxGasLimit,
+    },
+  };
+
   return (
-    <div className="view-quote">
```
Comment (adonesky1) [inline ui/pages/swaps/view-quote/view-quote.js:676]: @jpuri this diff is really hard to parse, do you mind explaining what's changed in the return block here?

**Review by adonesky1 — COMMENTED** [review]
Few questions/comments. Generally looking good!

File: ui/components/app/edit-gas-fee-button/edit-gas-fee-button.js (hunk 3 above)
Comment (jpuri) [reply ui/components/app/edit-gas-fee-button/edit-gas-fee-button.js:45]: `estimateUsed` is used at other placed also for example when updating transaction to get gas values, I would avoid changing it

File: ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.js (hunk 4 above)
Comment (jpuri) [reply ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.js:48]: by designs, need to check with Jake

File: ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.js (hunk 5 above)
Comment (jpuri) [reply ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.js:58]: aggressive = swap suggested for swaps, it works same just different label

File: ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js (hunk 7 above)
Comment (jpuri) [reply ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js:161]: I already added tooltip, let me remove this

File: ui/hooks/gasFeeInput/test-utils.js (hunk 9 above)
Comment (jpuri) [reply ui/hooks/gasFeeInput/test-utils.js:134]: This condition is not logically same, just that I am returning same value for tests

File: ui/hooks/gasFeeInput/useMaxFeePerGasInput.js (hunk 10 above)
Comment (jpuri) [reply ui/hooks/gasFeeInput/useMaxFeePerGasInput.js:50]: for swaps I found that values of gas are used from `getCustomMaxFeePerGas` and also if we update the gas value update is made to `getCustomMaxFeePerGas` not transaction

File: ui/hooks/gasFeeInput/useMaxFeePerGasInput.js (hunk 11 above)
Comment (jpuri) [reply ui/hooks/gasFeeInput/useMaxFeePerGasInput.js:67]: This is to ensure that null coalescing operator used below works well.

File: ui/hooks/gasFeeInput/useMaxFeePerGasInput.js (hunk 12 above)
Comment (jpuri) [reply ui/hooks/gasFeeInput/useMaxFeePerGasInput.js:134]: plz check above

File: ui/hooks/gasFeeInput/useMaxPriorityFeePerGasInput.js [hunk 16]
```diff
@@ -52,22 +59,29 @@ export function useMaxPriorityFeePerGasInput({
 
   const showFiat = useSelector(getShouldShowFiat);
 
-  const maxPriorityFeePerGasFromTransaction = supportsEIP1559
-    ? getMaxPriorityFeePerGasFromTransaction(transaction)
-    : 0;
+  let initialMaxMaxPriorityFeePerGas;
```
[the remark this answers is not in the document]
Comment (jpuri) [reply ui/hooks/gasFeeInput/useMaxPriorityFeePerGasInput.js:62]: Fixed

File: ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js [hunk 17]
```diff
@@ -117,22 +133,16 @@ const EditGasItem = ({ priorityLevel }) => {
...
+        <I18nValue messageKey={title} />
       </span>
       <span
         className={`edit-gas-item__time-estimate edit-gas-item__time-estimate-${priorityLevel}`}
       >
-        {minWaitTime
-          ? minWaitTime && toHumanReadableTime(t, minWaitTime)
-          : '--'}
+        {editGasMode !== EDIT_GAS_MODES.SWAPS && (
+          <>{minWaitTime ? toHumanReadableTime(t, minWaitTime) : '--'}</>
```
[the remark this answers is not in the document]
Comment (jpuri) [reply ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js:145]: Fixed

File: ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js (hunk 7 above)
Comment (jpuri) [reply ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js:161]: Fixed

File: ui/pages/confirm-transaction-base/gas-details-item/gas-details-item-title/gas-details-item-title.js (hunk 14 above)
Comment (jpuri) [reply ui/pages/confirm-transaction-base/gas-details-item/gas-details-item-title/gas-details-item-title.js]: fixed

File: ui/pages/swaps/view-quote/view-quote.js (hunk 15 above)
Comment (jpuri) [reply ui/pages/swaps/view-quote/view-quote.js:676]: The main changes in this are:

1. Wrapping jsx in providers
```
    <GasFeeContextProvider
      editGasMode={EDIT_GAS_MODES.SWAPS}
      minimumGasLimit={usedGasLimit}
      transaction={transaction}
    >
      <TransactionModalContextProvider captureEventEnabled={false}>
```

2. Adding modals:
```
            <EditGasFeePopover />
            <AdvancedGasFeePopover />
```

3. Passing supportEIP1559V2 to FeeCard component:
```
supportsEIP1559V2={
                  EIP_1559_V2_ENABLED && networkAndAccountSupports1559
                }
```

File: ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.js (hunk 5 above)
Comment (adonesky1) [reply ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.js:58]: I feel like it might be more clear to just have a `<EditGasItem priorityLevel={PRIORITY_LEVELS.SWAPS} />` here? But wouldn't block on it!

File: ui/hooks/gasFeeInput/useMaxFeePerGasInput.js (hunk 11 above)
Comment (adonesky1) [reply ui/hooks/gasFeeInput/useMaxFeePerGasInput.js:67]: Could you be a bit more specific? Not following which instance of null coalescing your talking about here? on new line 128? If so, that is using `maxFeePerGas` not `initalMaxFeePerGas`?

File: ui/hooks/gasFeeInput/useMaxPriorityFeePerGasInput.js [hunk 18]
```diff
@@ -34,13 +36,18 @@ const getMaxPriorityFeePerGasFromTransaction = (transaction) => {
...
-  supportsEIP1559V2,
+  editGasMode,
   estimateToUse,
   gasEstimateType,
   gasFeeEstimates,
   gasLimit,
+  supportsEIP1559V2,
   transaction,
 }) {
+  const swapCustomMaxPriorityFeePerGas = useSelector(
```
Comment (adonesky1) [inline ui/hooks/gasFeeInput/useMaxPriorityFeePerGasInput.js:47]: could we add a comment here explaining why we use the `getCustomMaxPriorityFeePerGas` for this value?

**Review by adonesky1 — DISMISSED** [review]
LGTM. Just a small non-blocking question left.

File: ui/hooks/gasFeeInput/useMaxPriorityFeePerGasInput.js (hunk 18 above)
Comment (jpuri) [reply ui/hooks/gasFeeInput/useMaxPriorityFeePerGasInput.js]: I found that this was redundant, I just removed.

File: ui/hooks/gasFeeInput/useMaxFeePerGasInput.js [hunk 19]
```diff
@@ -60,16 +53,9 @@ export function useMaxFeePerGasInput({
...
-  if (editGasMode === EDIT_GAS_MODES.SWAPS && supportsEIP1559V2) {
-    initialMaxFeePerGas = swapCustomMaxFeePerGas
-      ? Number(hexWEIToDecGWEI(swapCustomMaxFeePerGas))
-      : null;
-  } else {
-    initialMaxFeePerGas = supportsEIP1559
-      ? getMaxFeePerGasFromTransaction(transaction)
-      : 0;
-  }
+  const initialMaxFeePerGas = supportsEIP1559
```
Comment (adonesky1) [inline ui/hooks/gasFeeInput/useMaxFeePerGasInput.js:56]: Could you explain why we no longer have to use `swapCustomMaxFeePerGas` when `editGasMode === EDIT_GAS_MODES.SWAPS`?

Comment (jpuri) [reply ui/hooks/gasFeeInput/useMaxFeePerGasInput.js:56]: Yep this code I had added in this PR itself, but I was checking this with swaps team and found that this is not needed. Initialisation with high estimate should work for swaps.

I made this change below: https://github.com/MetaMask/metamask-extension/pull/12966/commits/f77752bfd0d563a0a007ac40b98d42f4522c0aaa#diff-d8d61bfbae4f1ef755fb241b4dc5aed4523d7317392c51bd3cbb0bfcd2e55961R70 to fix setting value to `0` and that worked well for initialising swaps correctly.

**Comment by darkwing** [comment]
@dan437 This one is important for you to review!

**Review by adonesky1 — DISMISSED** [review]

File: ui/hooks/gasFeeInput/useGasFeeInputs.test.js [hunk 20]
```diff
@@ -350,4 +351,13 @@ describe('useGasFeeInputs', () => {
       expect(result.current.supportsEIP1559V2).toBe(false);
     });
   });
+
+  describe('editGasMode', () => {
+    it('should return editGasMode passed', () => {
```
Comment (dan437) [inline ui/hooks/gasFeeInput/useGasFeeInputs.test.js:356]: Shouldn't it say something like: 'should return editGasMode for Swaps'?

File: ui/pages/swaps/view-quote/view-quote.js [hunk 21]
```diff
@@ -99,6 +103,10 @@ import CountdownTimer from '../countdown-timer';
 import SwapsFooter from '../swaps-footer';
 import ViewQuotePriceDifference from './view-quote-price-difference';
 
+// eslint-disable-next-line prefer-destructuring
+const EIP_1559_V2_ENABLED =
```
Comment (dan437) [inline ui/pages/swaps/view-quote/view-quote.js:107]: It would be good to have a selector for EIP 1559 v2, like we have for EIP 1559 inside `ducks/metamask/metamask.js`:
```
export function isEIP1559Network(state) {
  return state.metamask.networkDetails?.EIPS[1559] === true;
}
```

**Review by adonesky1 — DISMISSED** [review]

File: ui/pages/swaps/view-quote/view-quote.js [hunk 22]
```diff
@@ -672,93 +680,147 @@ export default function ViewQuote() {
...
+                <EditGasPopover
+                  transaction={transaction}
+                  minimumGasLimit={usedGasLimit}
+                  defaultEstimateToUse={GAS_RECOMMENDATIONS.HIGH}
+                  mode={EDIT_GAS_MODES.SWAPS}
+                  confirmButtonText={t('submit')}
+                  onClose={onCloseEditGasPopover}
+                />
+              )}
+            <EditGasFeePopover />
```
Comment (dan437) [inline ui/pages/swaps/view-quote/view-quote.js:728]: Don't we want to render EIP 1559 v2 components only if `EIP_1559_V2_ENABLED` is true? Because otherwise they would slow down rendering a little.

**Review by dan437 — COMMENTED** [review]
Amazing job! I've re-tested the latest fixes and it seems to be working well now in Swaps. Thank you for making the changes!

**Review by mcmire — DISMISSED** [review]

File: ui/pages/swaps/view-quote/view-quote.js (hunk 22 above)
Comment (jpuri) [reply ui/pages/swaps/view-quote/view-quote.js]: Good point, I made this change.

File: ui/pages/swaps/view-quote/view-quote.js (hunk 21 above)
Comment (jpuri) [reply ui/pages/swaps/view-quote/view-quote.js:107]: Yep but actually `EIP 1559 v2` is `isEIP1559Network(state) + feature_flag_enabled` so it is not purely a state thing.

File: ui/hooks/gasFeeInput/useGasFeeInputs.test.js (hunk 20 above)
Comment (jpuri) [reply ui/hooks/gasFeeInput/useGasFeeInputs.test.js:356]: This can actually be other gas modes also.

**Review by adonesky1 — APPROVED** [review]

File: ui/pages/swaps/view-quote/view-quote.js (hunk 21 above)
Comment (dan437) [reply ui/pages/swaps/view-quote/view-quote.js:107]: All that logic can be abstracted into a selector, so we don't need to duplicate this every time we need it:
```
const EIP_1559_V2_ENABLED = process.env.EIP_1559_V2 === true || process.env.EIP_1559_V2 === 'true';
const supportsEIP1559V2 = EIP_1559_V2_ENABLED && networkAndAccountSupports1559;
```

**Review by dan437 — APPROVED** [review]

File: ui/pages/swaps/view-quote/view-quote.js (hunk 21 above)
Comment (jpuri) [reply ui/pages/swaps/view-quote/view-quote.js:107]: I would actually avoid adding logic to selector, a separate hook or utility function might be actually a better place for it.
